### PR TITLE
Yossi/proxy generator

### PR DESCRIPTION
**Pull request by yossi-stratoscale**
Enhancing hammock so proxy endpoints can become generators to run code before and after the request is proxied.
@Stratoscale/mgmt @eyal-stratoscale Please review harshly.

[ Jenkins check-it Job underway: http://jenkins.dc1.strato:8080/view/MGMT/job/mgmt-check-it%20clean/571/ ]

### CODE REVIEW

File: tests/test_redirect.py [hunk 1]
```diff
@@ -1,13 +1,23 @@
 from __future__ import absolute_import
+
+import os
```
Comment (maroun-stratoscale) [inline tests/test_redirect.py]: unused import (and `six` as well).

**Comment by yossi-stratoscale** [comment]
@Stratoscale/mgmt ping

File: hammock/common.py [hunk 2]
```diff
@@ -60,8 +60,7 @@ def func_is_pass(func):
     while not lines.pop(0).startswith("def"):
         pass
     empty = "".join(lines).strip() == "pass"
-    if not empty:
-        raise Exception("Passthrough function %s is not empty", func.__name__)
+    return empty
```
Comment (eyal-stratoscale) [inline hammock/common.py]: :+1:
Question, would you like to check if the `func` is a generator here? As I understand, functions that are not generators are not valid here?

Comment (eyal-stratoscale) [inline hammock/common.py]: :+1:
Question, would you like to check if the func is a generator here? As I understand, functions that are not generators are not valid here?

File: hammock/wrappers/_route.py [hunk 3]
```diff
@@ -67,42 +71,59 @@ def _wrapper(self, req):
...
 
+        if self.dest is None:
             # Invoke the routing method:
             logging.debug('[kwargs %s] %s', req.uid, kwargs)
             result = self(**kwargs)
 
             resp = response.Response.from_result(result, self.success_code, self.response_content_type)
         else:
+            generator = self(**kwargs)
+            # code before 'yield'
```
Comment (eyal-stratoscale) [inline hammock/wrappers/_route.py]: can you make the comment more descriptive?

File: hammock/wrappers/_sink.py [hunk 4]
```diff
@@ -1,18 +1,36 @@
...
             credentials = self.credentials_class(req.headers)
             req.url_params[common.KW_CREDENTIALS] = credentials
 
         if self.dest is None:
             return self(req, **req.url_params)  # pylint: disable=not-callable
         else:
-            return proxy.proxy(req, self.dest)
+            generator = self(req, **req.url_params)
+            # code before 'yield'
+            next(generator)
```
Comment (eyal-stratoscale) [inline hammock/wrappers/_sink.py]: here it will be nice to do: `req = next(generator) or req`
so the generator can modify the request and return it

Comment (yossi-stratoscale) [inline hammock/wrappers/_sink.py]: can't it modify it by reference? (it gets it in its args)

File: hammock/common.py (hunk 2 above)
Comment (yossi-stratoscale) [inline hammock/common.py]: you suggest that for a generator the answer to func_is_pass is True? I don't think it's reasonable.
Yes, only generators and pass-only-funcs are valid.
(Maybe we can rename this utility function to assert_valid_proxy_func or is_valid_proxy_func)

File: hammock/wrappers/_sink.py (hunk 4 above)
Comment (eyal-stratoscale) [inline hammock/wrappers/_sink.py]: it can, i suppose you are right.

**Pull request merged**